### PR TITLE
Add Expect (SSH) workflows

### PR DESCRIPTION
**Pull request by nebilbenmabrouk**

### CODE REVIEW

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml [hunk 1]
```diff
@@ -0,0 +1,290 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<job
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns="urn:proactive:jobdescriptor:3.13" xsi:schemaLocation="urn:proactive:jobdescriptor:3.13 http://www.activeeon.com/public_content/schemas/proactive/jobdescriptor/3.13/schedulerjob.xsd"  name="Expect_Remote_Command_Execution" projectName="2. Advanced Workflows" priority="normal" onTaskError="continueJobExecution"  maxNumberOfExecution="2"  >
+  <variables>
+    <variable name="SSH_HOST" value="ssh.server.net" model="PA:NOT_EMPTY_STRING" description="Hostname or IP address of the remote SSH server machine" group="SSH Connection" advanced="false" hidden="false"/>
```
Comment (fviale) [inline GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:6]: there are too many variables at the job level and not enough at the task level.

This workflow is meant to be used as a drag/drop inside an existing workflow rather than a standalone one.

I understand that we may want to keep some variables at the job-level, but most variables should be at task level

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+          <pathElement path="expect4j-1.9.jar"/>
+          <pathElement path="com.springsource.tcl.lang-1.4.1.jar"/>
+          <pathElement path="com.springsource.tcl.lang.jacl-1.4.1.jar"/>
+          <pathElement path="org.apache.servicemix.bundles.oro-2.0.8_6.jar"/>
+        </additionalClasspath>
+      </forkEnvironment>
+      <pre>
+        <script>
+          <code language="bash">
+            <![CDATA[
```
Comment (fviale) [inline GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:57]: please add a comment saying that this script supports only single-line commands

File: GetStarted/resources/catalog/Expect_Remote_Script_Execution.xml [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+}
+
+try {   
+    cmdsToExecute = new ArrayList<String>();
+    localFile = new File(sshLocalFile)
+    binaryContent = localFile.bytes
+    base64Content = binaryContent.encodeBase64().toString()
+	cmdsToExecute.add("echo '" + base64Content + "' > " + sshRemoteFile)
+    cmdsToExecute.add("cat " + sshRemoteFile + " | base64 --decode > " + sshRemoteFile + ".sh")
+    cmdsToExecute.add("rm " + sshRemoteFile)
```
[the remark this answers is not in the document]
Comment (nebilbenmabrouk) [reply GetStarted/resources/catalog/Expect_Remote_Script_Execution.xml]: Done

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    <variable name="SSH_COMMAND" value="echo &#39;hello world&#39;"  description="The command to be executed on the SSH server machine" group="SSH Command" advanced="false" hidden="false"/>
+    <variable name="SSH_PROMPT_REGEXP" value="${SSH_USER}@[A-Za-z0-9_.-]+:.+$" model="PA:NOT_EMPTY_STRING" description="Regular expresion describing the set of CLI prompts in the SSH server system. The CLI prompts must be separated by a double pipe character ||" group="SSH Command" advanced="true" hidden="false"/>
+  </variables>
+  <description>
+    <![CDATA[ A workflow example that executes one or more shell commands on a remote machine using SSH protocol. It is based on the Expect4j tool to perform an interactive SSH authentication (using a password or a private key). The executed commands are provided in the task pre-script. Besides, the user can provide another SSH command in the SSH_COMMAND variable. This command will be appended at the end of the script. ]]>
```
[the remark this answers is not in the document]
Comment (nebilbenmabrouk) [reply GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml]: Done

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml (hunk 2 above)
Comment (nebilbenmabrouk) [reply GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:52]: Done

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml (hunk 1 above)
Comment (nebilbenmabrouk) [reply GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:6]: The fact considering a variable as a workflow variable or a task variable depends on the workflow functions.
If we consider a typical workflow composed of multiple SSH and Expect tasks executing various commands against a SSH server, we need to keep at the workflow level the variables that are common to all the tasks (e.g., host, port, creds, prompt). So I moved only the SSH_Command variable at the task level. Do you agree?

Comment (fviale) [reply GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:6]: For common variables that can be put at job level, it is important to duplicate them at task-level with inherited true.

This is easier to integrate in case the values of these variables need to be different for several tasks. In that case, the user simply need to delete the variables at job-level.

If you don't do this, then the user need to manually copy each variable name, value, model and description to each individual task.

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    <variable name="SSH_PROMPT_REGEXP" value="${SSH_USER}@[A-Za-z0-9_.-]+:.+$" model="PA:NOT_EMPTY_STRING" description="Regular expresion describing the set of CLI prompts in the SSH server system. The CLI prompts must be separated by a double pipe character ||" group="SSH Connection" advanced="true" hidden="false"/>
+  </variables>
+  <description>
+    <![CDATA[ A workflow example that executes one or more shell commands on a remote machine using SSH protocol. It is based on the Expect4j tool to perform an interactive SSH authentication (using a password or a private key). The executed commands are provided in the task pre-script, but the script must contain only single-line commands. Besides, the user can provide another SSH command in the SSH_COMMAND variable. This command will be appended at the end of the script. ]]>
```
Comment (fviale) [inline GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:16]: As there are two workflows with similar names, we should highlight the differences in the description. The single line "but the script must contain only single-line commands" is not enough to help users understand.

I would rephrase the paragraph with:

Executed commands are provided in the task pre-script and using the SSH_COMMAND variable.

There are two versions of this workflow:
 - Expect_Remote_Command_Execution (current) : the pre-script must contain only single-line commands. Each line will be executed one by one and interactively on the remote SSH server.
 - Expect_Remote_Script_Execution : the pre-script can be any bash script with multi-line commands. The script will be executed as a while on the remote SSH server.

On the other workflow description, simply switch (current).

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml (hunk 2 above)
Comment (fviale) [reply GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:52]: I was thinking of the bash comment in the beginning here
# this script can contain only single-line commands

File: GetStarted/resources/catalog/Expect_Remote_Script_Execution.xml [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+  <variables>
+    <variable name="SSH_HOST" value="ssh.server.net" model="PA:NOT_EMPTY_STRING" description="Hostname or IP address of the remote SSH server machine" group="SSH Connection" advanced="false" hidden="false"/>
+    <variable name="SSH_PORT" value="22" model="PA:INTEGER" description="The listening port of the SSH server. Default value is 22" group="SSH Connection" advanced="true" hidden="false"/>
+    <variable name="SSH_USER" value="my-ssh-login" model="PA:NOT_EMPTY_STRING" description="SSH user name" group="SSH Connection" advanced="false" hidden="false"/>
+    <variable name="SSH_CREDENTIAL_TYPE" value="SSH_PASSWORD" model="PA:LIST(SSH_PASSWORD,SSH_PRIVATE_KEY)" description="SSH authentication type (password or private key)" group="SSH Connection" advanced="false" hidden="false"/>
+    <variable name="SSH_PASSWORD" value="ENC(fGGsmfrb4fQTPz9MP/UQ6Q==)" model="PA:HIDDEN" description="SSH server password" group="SSH Connection" advanced="false" hidden="false"/>
```
Comment (fviale) [inline GetStarted/resources/catalog/Expect_Remote_Script_Execution.xml:10]: could you please let the value empty instead of having a default value ?

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml (hunk 5 above)
Comment (nebilbenmabrouk) [reply GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml]: Done.

File: GetStarted/resources/catalog/Expect_Remote_Script_Execution.xml (hunk 6 above)
Comment (nebilbenmabrouk) [reply GetStarted/resources/catalog/Expect_Remote_Script_Execution.xml:10]: Done.

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+  <variables>
+    <variable name="SSH_HOST" value="ssh.server.net" model="PA:NOT_EMPTY_STRING" description="Hostname or IP address of the remote SSH server machine" group="SSH Connection" advanced="false" hidden="false"/>
+    <variable name="SSH_PORT" value="22" model="PA:INTEGER" description="The listening port of the SSH server. Default value is 22" group="SSH Connection" advanced="true" hidden="false"/>
+    <variable name="SSH_USER" value="my-ssh-login" model="PA:NOT_EMPTY_STRING" description="SSH user name" group="SSH Connection" advanced="false" hidden="false"/>
+    <variable name="SSH_CREDENTIAL_TYPE" value="SSH_PASSWORD" model="PA:LIST(SSH_PASSWORD,SSH_PRIVATE_KEY)" description="SSH authentication type (password or private key)" group="SSH Connection" advanced="false" hidden="false"/>
+    <variable name="SSH_PASSWORD" value="ENC(nZHfe4xn2g2fbw7cY4b2rg==)" model="PA:HIDDEN" description="SSH server password" group="SSH Connection" advanced="false" hidden="false"/>
```
[the remark this answers is not in the document]
Comment (nebilbenmabrouk) [reply GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:10]: Done.

**Review by fviale — APPROVED** [review]
LGTM
Don't forget to squash before merging

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml [hunk 8]
```diff
@@ -0,0 +1,298 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<job
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xmlns="urn:proactive:jobdescriptor:3.13" xsi:schemaLocation="urn:proactive:jobdescriptor:3.13 http://www.activeeon.com/public_content/schemas/proactive/jobdescriptor/3.13/schedulerjob.xsd"  name="Expect_Remote_Command_Execution" projectName="2. Advanced Workflows" priority="normal" onTaskError="continueJobExecution"  maxNumberOfExecution="2"  >
+  <variables>
+    <variable name="SSH_HOST" value="ssh.server.net" model="PA:NOT_EMPTY_STRING" description="Hostname or IP address of the remote SSH server machine" group="SSH Connection" advanced="false" hidden="false"/>
+    <variable name="SSH_PORT" value="22" model="PA:INTEGER" description="The listening port of the SSH server. Default value is 22" group="SSH Connection" advanced="true" hidden="false"/>
```
Comment (mklkun) [inline GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:7]: Since all variables are related to the same group `SSH Connection` Why calling them `SSH_*`?

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        <variable name="SSH_PROMPT_REGEXP" value="${SSH_USER}@[A-Za-z0-9_.-]+:.+$" model="PA:NOT_EMPTY_STRING" description="Regular expression describing the set of CLI prompts in the SSH server system. The CLI prompts must be separated by a double pipe character ||" inherited="true" advanced="true" hidden="false"/>
+        <variable name="SSH_COMMAND" value="echo &#39;hello world&#39;" description="The command to be executed on the SSH server machine" inherited="false" advanced="false"  hidden="false"/>
+        <variable name="DEBUG" value="false" description="Enable the debug mode" inherited="false" model="PA:BOOLEAN" advanced="false" hidden="false"/>
+      </variables>
+      <genericInformation>
+        <info name="task.icon" value="/automation-dashboard/styles/patterns/img/wf-icons/ssh.png"/>
+        <info name="PRE_SCRIPT_AS_FILE" value="script.sh"/>
```
Comment (mklkun) [inline GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:45]: I don't know if it is relevant here but could we have a more relevant file name here rather than `script.sh`?

File: GetStarted/resources/catalog/Expect_Remote_Script_Execution.xml [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+    <variable name="SSH_PASSWORD" value="" model="PA:HIDDEN" description="SSH server password" group="SSH Connection" advanced="false" hidden="false"/>
+    <variable name="SSH_PRIVATE_KEY" value="${SSH_USER}@${SSH_HOST}" model="PA:CREDENTIAL" description="SSH private key" group="SSH Connection" advanced="false" hidden="true"/>
+    <variable name="SSH_CREDENTIAL_HANDLER" value="" model="PA:SPEL(variables[&#39;SSH_CREDENTIAL_TYPE&#39;] == &#39;SSH_PASSWORD&#39; ? showVar(&#39;SSH_PASSWORD&#39;) &amp;&amp; hideVar(&#39;SSH_PRIVATE_KEY&#39;) : showVar(&#39;SSH_PRIVATE_KEY&#39;) &amp;&amp; hideVar(&#39;SSH_PASSWORD&#39;))" description="SSH credential handler that shows and displays dynamically variables depending on the authentication type" group="SSH Connection" advanced="false" hidden="true"/>
+    <variable name="SSH_PROMPT_REGEXP" value="${SSH_USER}@[A-Za-z0-9_.-]+:.+$" model="PA:NOT_EMPTY_STRING" description="Regular expresion describing the set of CLI prompts in the SSH server system. The CLI prompts must be separated by a double pipe character ||" group="SSH Connection" advanced="true" hidden="false"/>
```
Comment (mklkun) [inline GetStarted/resources/catalog/Expect_Remote_Script_Execution.xml:13]: Same for these variables naming.

**Review by mklkun — APPROVED** [review]
2 not blocking comments.

File: GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml (hunk 8 above)
Comment (nebilbenmabrouk) [reply GetStarted/resources/catalog/Expect_Remote_Command_Execution.xml:7]: To be more explicit for the users (even if they do not pay attention to the group, they understand the variables)
